incubator/hnc/pkg/controllers/object_controller_new.go

```diff
@@ -38,25 +60,307 @@ type ObjectReconcilerNew struct {
+	if srcInst == nil {
+		return remove, nil
+	}
+
+	// If the copy is different from the source, return the write action and the source instance.
+	if !reflect.DeepEqual(object.Canonical(inst), object.Canonical(srcInst)) {
+		metadata.SetLabel(inst, api.LabelInheritedFrom, srcInst.GetNamespace())
+		return write, srcInst
+	}
+
+	return ignore, nil
+}
+
+// syncSource syncs the copy in the forest with the current source object. If there's a change,
+// enqueue all the descendants to propagate the new source.
+func (r *ObjectReconcilerNew) syncSource(ctx context.Context, log logr.Logger, src *unstructured.Unstructured) {
+	// syncPropagated doesn't check isExcluded because the propagation won't happen if it's excluded.
+	if r.isExcluded(log, src) {
+		return
+	}
+	sns := src.GetNamespace()
+	n := src.GetName()
+	origCopy := r.Forest.Get(sns).GetOriginalObject(r.GVK, n)
+
+	// Early exit if the source object exists and remains unchanged.
+	if origCopy != nil && reflect.DeepEqual(object.Canonical(src), object.Canonical(origCopy)) {
+		log.Info("Unchanged Source")
+		return
+	}
+
+	// Update or create a copy of the source object in the forest
+	r.Forest.Get(sns).SetOriginalObject(src.DeepCopy())
+
+	// Enqueue all the descendant copies
+	r.enqueueDescendants(ctx, log, src)
+}
+
+func (r *ObjectReconcilerNew) enqueueDescendants(ctx context.Context, log logr.Logger, src *unstructured.Unstructured) {
+	sns := src.GetNamespace()
+	dns := r.Forest.Get(sns).DescendantNames()
+	for _, ns := range dns {
+		dc := object.Canonical(src)
+		dc.SetNamespace(ns)
+		log.Info("Enqueuing descendant copy", "affected", ns+"/"+src.GetName(), "reason", "The source changed")
+		r.Affected <- event.GenericEvent{Meta: dc}
+	}
+}
+
+// enqueuePropagatedObjects is only called from SyncNamespace. It's the only place a forest lock is
+// needed in SyncNamespace, so we made it into a function with forest lock instead of holding the
+// lock for the entire SyncNamespace.
+func (r *ObjectReconcilerNew) enqueuePropagatedObjects(ctx context.Context, log logr.Logger, ns string) {
+	r.Forest.Lock()
+	defer r.Forest.Unlock()
+
+	// Enqueue local copies of the original objects in the ancestors from forest.
+	o := r.Forest.Get(ns).GetPropagatedObjects(r.GVK)
+	for _, obj := range o {
+		lc := object.Canonical(obj)
+		lc.SetNamespace(ns)
+		log.Info("Enqueuing local copy of the ancestor original for reconciliation", "affected", lc.GetName())
+		r.Affected <- event.GenericEvent{Meta: lc}
+	}
+}
+
+// operate operates the action generated from syncing the object with the forest.
+func (r *ObjectReconcilerNew) operate(ctx context.Context, log logr.Logger, act action, inst, srcInst *unstructured.Unstructured) error {
+	switch act {
+	case ignore:
+		return nil
+	case remove:
+		return r.delete(ctx, log, inst)
+	case write:
+		return r.write(ctx, log, inst, srcInst)
+	default:
+		// Generate log for any unset action.
+		log.Info("ACTION UNSET!!")
```

Comment:
Shouldn't this be an error? Although we should never actually hit this, should we really suppress/ignore this if it does happen? If this is really the best choice, some comments on why it's ok to ignore would be helpful.

Reply:
+1 to making this an error